lib/ex_grok/master_supervisor.ex

```diff
@@ -0,0 +1,17 @@
+defmodule ExGrok.MasterSupervisor do
+  @moduledoc false
+  use Supervisor
+
+  def start_link do
+    Supervisor.start_link(__MODULE__, [], name: :master_supervisor)
+  end
+
+  def init(_) do
+    children = [
+      worker(ExGrok.Ngrok, []),
```

Comment:
What it deals with here is actually the other way around - once the `ngrok` executable dies for some reason (you can try killing it in terminal using its pid - `port` health_check is performed every 3 sec so this will be detected) then new port will be opened.

In the gen_server itself there is very little that can go wrong, effectively causing the error taking it down (and as such being restarted by supervisor). I'm choosing not to go with the _wrapper script_ just yet - I want to understand better how things work, where the limits are, what possible problems can be (maybe I'm totally wrong with my assumption that _very little can go wrong_) and reach out for the solution in last resort.

From my current observation it looks like when shutting application down, the `ngrok` executable dies as well, though, not immediately.

I'm looking into it with much interest!